### PR TITLE
jamesSid/AutoScore

### PR DESCRIPTION
**Pull request by voidSevenSevenSix**

### CODE REVIEW

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import com.pathplanner.lib.PathConstraints;
+import com.pathplanner.lib.PathPlanner;
+import com.pathplanner.lib.PathPlannerTrajectory;
+import com.pathplanner.lib.commands.FollowPathWithEvents;
+
+import frc.robot.Cal;
+import frc.robot.subsystems.drive.DriveSubsystem;
+
+public class AutoScore extends SequentialCommandGroup{
+    private DriveSubsystem drive;
```
Comment (balghane) [inline src/main/java/frc/robot/commands/AutoScore.java]: Not needed?

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        new PathConstraints(Cal.PLACEHOLDER_DOUBLE, Cal.PLACEHOLDER_DOUBLE));
+    private HashMap<String, Command> eventMap = new HashMap<>();
+    private boolean isFirstPath = true;
+    
+    public AutoScore(Lift lift, Intake intake, DriveSubsystem drive, TagLimelight tagLimelight){
+        eventMap.put("deployIntake", new InstantCommand(()->{intake.setDesiredDeployed(true);}, intake));
+        eventMap.put("closeIntake", new InstantCommand(()->{intake.setDesiredDeployed(false);}, intake));
+        addCommands(
+            new InstantCommand(lift::manualPrepScore, lift),
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.MANUAL_PREP_SCORE)),
```
Comment (balghane) [inline src/main/java/frc/robot/commands/AutoScore.java]: This unfortunately isn't a position, I think Asher has the correct position in a PR that just merged

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.SCORE_HIGH_CONE)),
+            new finishScore(),
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.STARTING)),
+            new FollowPathWithEvents(
+                drive.followTrajectoryCommand(trajInit, true),
+                trajInit.getMarkers(), eventMap),
+            /*limelight goes here */
+            new InstantCommand(lift::manualPrepScore, lift),
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.MANUAL_PREP_SCORE)),
+            drive.followTrajectoryCommand(trajCharge, isFirstPath), //this does not accept the FollowPathWithEvents
```
Comment (balghane) [inline src/main/java/frc/robot/commands/AutoScore.java]: This time, isFirstPath is false

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        PathPlanner.loadPath(
+    "ScoringLocToChargeStation",
+        new PathConstraints(Cal.PLACEHOLDER_DOUBLE, Cal.PLACEHOLDER_DOUBLE));
+    private HashMap<String, Command> eventMap = new HashMap<>();
+    private boolean isFirstPath = true;
+    
+    public AutoScore(Lift lift, Intake intake, DriveSubsystem drive, TagLimelight tagLimelight){
+        eventMap.put("deployIntake", new InstantCommand(()->{intake.setDesiredDeployed(true);}, intake));
+        eventMap.put("closeIntake", new InstantCommand(()->{intake.setDesiredDeployed(false);}, intake));
+        addCommands(
```
Comment (balghane) [inline src/main/java/frc/robot/commands/AutoScore.java]: Need to addRequirements as well

**Review by balghane — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/AutoScore.java (hunk 2 above)
Comment (balghane) [reply src/main/java/frc/robot/commands/AutoScore.java]: So you can just fetch and rebase onto origin/main

**Comment by balghane** [comment]
Also better PR title and description pls?

**Comment by Panda8188** [comment]
> Also better PR title and description pls?

Yeah, James is crazy and does all the GitHub through the terminal instead of GitHub Desktop hahaha. Not sure if there is a way to give an actual description through the terminal though...

File: src/main/java/frc/robot/commands/AutoScore.java (hunk 2 above)
Comment (Panda8188) [reply src/main/java/frc/robot/commands/AutoScore.java]: wdym by this? Just a little confused by this

Comment (jcgh7) [reply src/main/java/frc/robot/commands/AutoScore.java]: he means that there is a position for manual prep score that asher pushed recently, and it should be replaced with that

**Comment by voidSevenSevenSix** [comment]
bruh he means for the pr

**Comment by Panda8188** [comment]
yeah but in github desktop it allows you to write a lot more with description and title and stuff than in terminal

File: src/main/java/frc/robot/commands/AutoScore.java (hunk 2 above)
Comment (Panda8188) [reply src/main/java/frc/robot/commands/AutoScore.java]: Ohhh I see

**Comment by Panda8188** [comment]
I wasn't sure what you meant by the "correct position" that Asher made, Basel. So, I just made the according changes for everything else.

**Comment by voidSevenSevenSix** [comment]
bro the pull request is on github.com are you talking about commiting?

**Comment by Panda8188** [comment]
oop yeah commiting

**Comment by tails618** [comment]
> I wasn't sure what you meant by the "correct position" that Asher made, Basel. So, I just made the according changes for everything else.

If you fetch+rebase onto main, you can look at the LiftPosition enum and see that I've added the position that's meant to be there. I can help you with rebasing if you want.

**Comment by Panda8188** [comment]
Yes please that would be greatly appreciated!

**Comment by tails618** [comment]
@Panda8188 first switch to the main branch and make sure you have the latest code pulled. Then switch back to your branch, and follow [these instructions](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch) (go to the rebase section), rebasing your branch onto main. (Not the other way around! I don't know if that's even possible.) There may be rebase conflicts, unfortunately. You may attempt to deal with those or I can look at them at some point.

**Comment by Panda8188** [comment]
After I do that, should I push whatever changes were made onto this branch? Or do I just leave the code as it is?

**Comment by Panda8188** [comment]
Finally, all merge conflicts have been resolved and all requested changes from Basel have been made accordingly.

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+         /** Events include: open intake and close intake before and after obtaining game piece */
+        eventMap.put("deployIntake", new InstantCommand(()->{intake.setDesiredDeployed(true);}, intake));
+        eventMap.put("closeIntake", new InstantCommand(()->{intake.setDesiredDeployed(false);}, intake));
+        /** TODO: configure limelight to "take over" driving process after certain point AFTER obtaining game piece to be more accurate with distance */
+
+
+        /** Initialize sequential commands that run for the "15 second autonomous phase" */
+        addCommands(
+            new InstantCommand(lift::ManualPrepScoreSequence, lift),
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.MANUAL_PREP_SCORE)),
```
Comment (avakimc) [inline src/main/java/frc/robot/commands/AutoScore.java]: What does MANUAL_PREP_SCORE position mean? Like what does that do for the lift?

Comment (avakimc) [reply src/main/java/frc/robot/commands/AutoScore.java]: That should not be there. It should be SCORE_HIGH_CONE

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            new finishScore(lift),
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.STARTING)),
+            new FollowPathWithEvents(
+                drive.followTrajectoryCommand(trajInit, true),
+                trajInit.getMarkers(), eventMap),
+
+            /** TODO: Limelight code goes here */
+
+            new InstantCommand(lift::ManualPrepScoreSequence, lift),
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.MANUAL_PREP_SCORE)),
```
Comment (avakimc) [inline src/main/java/frc/robot/commands/AutoScore.java]: this should also be SCORE_HIGH_CONE

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        eventMap.put("deployIntake", new InstantCommand(()->{intake.setDesiredDeployed(true);}, intake));
+        eventMap.put("closeIntake", new InstantCommand(()->{intake.setDesiredDeployed(false);}, intake));
+        /** TODO: configure limelight to "take over" driving process after certain point AFTER obtaining game piece to be more accurate with distance */
+
+
+        /** Initialize sequential commands that run for the "15 second autonomous phase" */
+        addCommands(
+            new InstantCommand(lift::ManualPrepScoreSequence, lift),
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.MANUAL_PREP_SCORE)),
+            new InstantCommand(lift::startScore, lift),
```
Comment (avakimc) [inline src/main/java/frc/robot/commands/AutoScore.java]: startScore is incorrect. It should be:
if (scoreLoc.isCone())
not 
if (!scoreLoc.isCone())

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.SCORE_HIGH_CONE)),
+            new finishScore(lift),
```
Comment (avakimc) [inline src/main/java/frc/robot/commands/AutoScore.java]: finishScore should close the grabber as well

**Review by avakimc — CHANGES_REQUESTED** [review]
*Change MANUAL_PREP-SCORE to SCORE_HIGH_CONE (or whatever it is called)
*startScore and finishScore methods need changes, but not in this PR

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    "InitScoreAndGetGamePiece",
+        new PathConstraints(Cal.PLACEHOLDER_DOUBLE, Cal.PLACEHOLDER_DOUBLE));
+
+    private PathPlannerTrajectory trajCharge =
+        PathPlanner.loadPath(
+    "ScoringLocToChargeStation",
+        new PathConstraints(Cal.PLACEHOLDER_DOUBLE, Cal.PLACEHOLDER_DOUBLE));
+
+    private HashMap<String, Command> eventMap = new HashMap<>();
+    private boolean isFirstPath = false;
```
Comment (balghane) [inline src/main/java/frc/robot/commands/AutoScore.java]: Do we need a variable for this?

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.STARTING)),
+            new FollowPathWithEvents(
+                drive.followTrajectoryCommand(trajInit, true),
+                trajInit.getMarkers(), eventMap),
+
+            /** TODO: Limelight code goes here */
+
+            new InstantCommand(lift::ManualPrepScoreSequence, lift),
+            new WaitUntilCommand(()->lift.atPosition(LiftPosition.MANUAL_PREP_SCORE)),
+            drive.followTrajectoryCommand(trajCharge, isFirstPath), //this does not accept the FollowPathWithEvents
```
Comment (balghane) [inline src/main/java/frc/robot/commands/AutoScore.java]: Are we ready to move onto the charge station at this point?

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 11]
```diff
@@ -46,7 +46,8 @@ public enum LiftPosition {
     PRE_SCORE_HIGH_CONE,
     POST_SCORE_HIGH,
     OUTTAKING,
-    STARTING
+    STARTING,
+    MANUAL_PREP_SCORE
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java]: Remove?

File: src/main/java/frc/robot/commands/AutoScore.java (hunk 6 above)
Comment (balghane) [reply src/main/java/frc/robot/commands/AutoScore.java]: Doesn't the prep score sequence go to PRE_SCORE_HIGH_CONE ?

**Review by balghane — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/AutoScore.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    public AutoScore(Lift lift, Intake intake, DriveSubsystem drive, TagLimelight tagLimelight){
+        addRequirements(lift, intake, drive, tagLimelight);
+         /** Events include: open intake and close intake before and after obtaining game piece */
+        eventMap.put("deployIntake", new InstantCommand(()->{intake.setDesiredDeployed(true);}, intake));
+        eventMap.put("closeIntake", new InstantCommand(()->{intake.setDesiredDeployed(false);}, intake));
+        /** TODO: configure limelight to "take over" driving process after certain point AFTER obtaining game piece to be more accurate with distance */
+
+
+        /** Initialize sequential commands that run for the "15 second autonomous phase" */
+        addCommands(
```
Comment (balghane) [inline src/main/java/frc/robot/commands/AutoScore.java]: We probably need to set the desired scoring locations in the ScoringUtil

File: src/main/java/frc/robot/commands/AutoScore.java (hunk 9 above)
Comment (jcgh7) [reply src/main/java/frc/robot/commands/AutoScore.java]: no i got rid of it

File: src/main/java/frc/robot/commands/AutoScore.java (hunk 10 above)
Comment (jcgh7) [reply src/main/java/frc/robot/commands/AutoScore.java]: i am not sure

File: src/main/java/frc/robot/commands/AutoScore.java (hunk 12 above)
Comment (jcgh7) [reply src/main/java/frc/robot/commands/AutoScore.java]: you mean preset the autonomous scoring locations?

**Review by balghane — APPROVED** [review]

**Pull request merged**